### PR TITLE
Requirements dev file

### PR DESCRIPTION
**Pull request by oddbookworm**
Continuation of #2698, using a branch on this repo instead of my own

### CODE REVIEW

File: requirements-dev.txt [hunk 1]
```diff
@@ -0,0 +1,7 @@
+black<=23.7.0
+clang-format<=17.0.6
+numpy<=1.26.3
+opencv-python<=4.8.1.78
```
Comment (ankith26) [inline requirements-dev.txt:4]: numpy and opencv aren't really dev dependencies, they're more like optional runtime dependencies

**Review by ankith26 — COMMENTED** [review]
I'm not sure if CI should be using this requirements dev file? It negatively affects build times, and in some builds this has an extreme effect.

**Comment by oddbookworm** [comment]
> I'm not sure if CI should be using this requirements dev file? It negatively affects build times, and in some builds this has an extreme effect.

I think that's part of the point of having one. If we pin the dep versions, then when something changes that would affect us, the CI would use the new version and would break. Unless we manually pin the versions in the CI too, that wouldn't really fix anything.

File: requirements-dev.txt (hunk 1 above)
Comment (Starbuck5) [reply requirements-dev.txt:4]: It also looks like numpy and opencv failed builds on pypy are the cause of most? all? of the current CI failures.

What I'm thinking right now is that opencv should be dropped from here and numpy should be hard pinned to the last version that ships pypy3.8 wheels (1.24.4)

We test in CI with numpy, we don't test with opencv.

Comment (oddbookworm) [reply requirements-dev.txt]: Can't pin to 1.24.4 numpy because it didn't ship with 3.12 wheels

Comment (MyreMylar) [reply requirements-dev.txt]: Probably if we are going to use numpy as a dependency we should synch to their 3.5 years/42 month community python version deprecation schedule as well or we will keep having these out of synch issues. Their schedule is defined in this NEP:

https://numpy.org/neps/nep-0029-deprecation_policy.html

3.8 was deprecated on April 14th 2023 and 3.9 will be deprecated this April.

Comment (Starbuck5) [reply requirements-dev.txt]: I'd rather keep a bit more extended support, as we've been doing.

I guess we should not pin numpy in the requirements file.

Also, numpy 2 is coming out if y'all haven't heard.

**Comment by narilee2006** [comment]
Why not requirements.txt?

**Comment by Starbuck5** [comment]
Requirements-dev is a common way to express development dependencies, I think. These aren't dependencies for pygame-ce's runtime, they are dependencies for building/testing pygame-ce.

**Comment by Starbuck5** [comment]
Does this need to be updated for/because of https://github.com/pygame-community/pygame-ce/pull/2744 ?

**Comment by oddbookworm** [comment]
Yup, about to push a commit updating

**Review by MyreMylar — APPROVED** [review]
LGTM 👍

**Comment by oddbookworm** [comment]
I'll hold off on this until #2824 is resolved, needs updates from meson build too

File: requirements-dev.txt [hunk 2]
```diff
@@ -0,0 +1,5 @@
+ruff<=0.4.4
+clang-format<=18.1.1
```
Comment (ankith26) [inline requirements-dev.txt:2]: In my understanding, these won't be dependencies, as pre-commit automatically takes care of it behind the scenes.

File: requirements-dev.txt [hunk 3]
```diff
@@ -0,0 +1,5 @@
...
+numpy
+pylint<=3.1.0
+Sphinx<=7.2.6
```
Comment (ankith26) [inline requirements-dev.txt:5]: Also, if #2853 is merged this line won't be needed, as the Sphinx dependency will be tracked in `pyproject.toml` along with other build-time dependencies.

File: requirements-dev.txt (hunk 1 above)
Comment (ankith26) [reply requirements-dev.txt]: What would be nice is to pass the equivalent of `--prefer-binary` specifically on the numpy dependency (and also maybe even opencv if we are adding that). Should fix our issues ig, idk

File: requirements-dev.txt [hunk 4]
```diff
@@ -0,0 +1,3 @@
+numpy
+pylint<=3.1.0
+meson
```
Comment (ankith26) [inline requirements-dev.txt:3]: meson is already covered due to specifying `meson-python` in `[build-system] requires` of `pyproject.toml`, so like the rest it need not be re-specified here.

However, you could specify `mypy` here instead, it's used by the stubcheck program we have

File: .github/workflows/build-windows.yml [hunk 5]
```diff
@@ -1,4 +1,3 @@
-name: Windows
```
Comment (ankith26) [inline .github/workflows/build-windows.yml:1]: why was this line removed

Comment (oddbookworm) [reply .github/workflows/build-windows.yml:1]: Oops, must've been a wonky merge conflict fix

**Comment by ankith26** [comment]
This PR has been superseded by `dev.py` and therefore I am closing this. But still, thanks for working on it as the work done here still inspired some design and behaviour of the current `dev.py`.
